File: solutions/java/Semaphore.java
// @include
public class Semaphore {
  private final int MAX_AVAILABLE;
  private int taken;

  public Semaphore(int maxAvailable) {
    this.MAX_AVAILABLE = maxAvailable;
    this.taken = 0;
  }

  public synchronized void acquire() throws InterruptedException {
    while (this.taken == MAX_AVAILABLE) {
      wait();
    }
    this.taken++;
  }

  public synchronized void release() throws InterruptedException {
    this.taken--;
    this.notifyAll();
  }
  // @exclude
  public int numTaken() { return taken; }
  // @include
}
// @exclude

class SemaphoreTest {
  static int sharedData = 0;

  public static void main(String[] args) {
    Semaphore semaphore = new Semaphore(1);
    System.out.println("Semaphore with 1 permit has been created");

    IncrementThread incrementThread1 = new IncrementThread(semaphore);
    new Thread(incrementThread1).start();

    DecrementThread decrementThread1 = new DecrementThread(semaphore);
    new Thread(decrementThread1).start();

    IncrementThread incrementThread2 = new IncrementThread(semaphore);
    new Thread(incrementThread2).start();

    DecrementThread decrementThread2 = new DecrementThread(semaphore);
    new Thread(decrementThread2).start();
  }
}

class IncrementThread implements Runnable {
  Semaphore semaphore;

  public IncrementThread(Semaphore semaphore) { this.semaphore = semaphore; }

  public void run() {
    while (true) {
      System.out.println(Thread.currentThread().getName()
                         + " is waiting for permit");
      try {
        semaphore.acquire();
        System.out.println(Thread.currentThread().getName()
                           + " has got permit, new count is "
                           + semaphore.numTaken());

        for (int i = 0; i < 3; i++) {
          Thread.sleep(200);
          System.out.println(Thread.currentThread().getName()
                             + " increments shared data: "
                             + SemaphoreTest.sharedData++);
        }

      } catch (InterruptedException e) {
        e.printStackTrace();
      }

      System.out.println(Thread.currentThread().getName()
                         + " has released permit");
      try {
        semaphore.release();
      } catch (InterruptedException e) {
        e.printStackTrace();
      }
    }
  }
}

class DecrementThread implements Runnable {
  Semaphore semaphore;

  public DecrementThread(Semaphore semaphore) { this.semaphore = semaphore; }

  public void run() {
    while (true) {
      System.out.println(Thread.currentThread().getName()
                         + " is waiting for permit");

      try {
        semaphore.acquire();
        System.out.println(Thread.currentThread().getName()
                           + " has got permit, new count is "
                           + semaphore.numTaken());

        for (int i = 0; i < 4; i++) {
          Thread.sleep(300);
          System.out.println(Thread.currentThread().getName()
                             + " is decrementing sharedData: "
                             + SemaphoreTest.sharedData--);
        }

      } catch (InterruptedException e) {
        e.printStackTrace();
      }

      System.out.println(Thread.currentThread().getName()
                         + " has released permit");
      try {
        semaphore.release();
      } catch (InterruptedException e) {
        e.printStackTrace();
      }
    }
  }
}

File: static/solutions/Semaphore.cc
#include <chrono>
#include <condition_variable>
#include <functional>
#include <iostream>
#include <mutex>
#include <string>
#include <thread>

using std::ref;
using std::string;
using std::cout;
using std::endl;
using std::chrono::milliseconds;
using std::thread;
using std::this_thread::sleep_for;
using std::mutex;
using std::lock_guard;
using std::unique_lock;
using std::condition_variable;

// @include
class Semaphore {
 public:
  Semaphore(int max_available) : max_available_(max_available), taken_(0) {}

  void Acquire() {
    unique_lock<mutex> lock(mx_);
    while (taken_ == max_available_) {
      cond_.wait(lock);
    }
    ++taken_;
  }

  void Release() {
    lock_guard<mutex> lock(mx_);
    --taken_;
    cond_.notify_all();
  }

  // @exclude
  int NumTaken() { return taken_; }
  // @include
 private:
  int max_available_;
  int taken_;
  mutex mx_;
  condition_variable cond_;
};
// @exclude

static int shared_data = 0;

void IncrementThread(Semaphore& semaphore, string name) {
  while (true) {
    cout << name << " is waiting for permit" << endl;
    semaphore.Acquire();
    cout << name << " has got permit, new count is " << semaphore.NumTaken()
         << endl;
    for (int i = 0; i < 3; ++i) {
      sleep_for(milliseconds(200));
      cout << name + " increments shared data: " << shared_data++ << endl;
    }
    cout << name + " has released permit" << endl;
    semaphore.Release();
  }
}

void DecrementThread(Semaphore& semaphore, string name) {
  while (true) {
    cout << name << " is waiting for permit" << endl;
    semaphore.Acquire();
    cout << name << " has got permit, new count is " << semaphore.NumTaken()
         << endl;
    for (int i = 0; i < 4; ++i) {
      sleep_for(milliseconds(300));
      cout << name + " decrements shared data: " << shared_data-- << endl;
    }
    cout << name + " has released permit" << endl;
    semaphore.Release();
  }
}

int main(int argc, char* argv[]) {
  Semaphore semaphore(1);
  cout << "Semaphore with 1 permit has been created" << endl;
  thread increment_thread1(IncrementThread, ref(semaphore),
                           "increment_thread_1");
  thread decrement_thread1(DecrementThread, ref(semaphore),
                           "decrement_thread_1");
  thread increment_thread2(IncrementThread, ref(semaphore),
                           "increment_thread_2");
  thread decrement_thread2(DecrementThread, ref(semaphore),
                           "decrement_thread_2");
  increment_thread1.join();
}
